proposals/non-trailing-named-arguments.md

```diff
@@ -0,0 +1,82 @@
+That is currently disallowed (`CS1738`) unless all the following arguments are also named.
+
+```C#
+DoSomething(isEmployed:true, name, age); // currently disallowed, even though all arguments are in position
+// CS1738 "Named argument specifications must appear after all fixed arguments have been specified"
+```
+
+Some additional examples:
+```C#
+public void DoSomething(bool isEmployed, string personName, int personAge) { ... }
+
+DoSomething(isEmployed:true, name, age); // currently CS1738, but would become legal
+DoSomething(true, personName:name, age); // currently CS1738, but would become legal
+DoSomething(name, isEmployed:true, age); // remains illegal
+DoSomething(name, age, isEmployed:true); // remains illegal
+DoSomething(true, personAge:age, personName:name); // already legal
+```
+
+This would also work with params:
+```C#
+public class Task
+{
+    public static Task When(TaskStatus all, TaskStatus any, params Task[] tasks);
+}
+Task.When(all: TaskStatus.RanToCompletion, any: TaskStatus.Faulted, task1, task2)
+```
+
+## Detailed design (C#)
+[design]: #detailed-design
+
+In §7.5.1 (Argument lists), the spec currently says:
+> An *argument* with an *argument-name* is referred to as a __named argument__, whereas an *argument* without an *argument-name* is a __positional argument__. It is an error for a positional argument to appear after a named argument in an *argument-list*.
+
+The proposal is to remove this error and update the rules for checking if a function member is applicable (§7.5.3.1):
+
+A function member is said to be an applicable function member with respect to an argument list A when all of the following are true:
+* [... existing rules ...]
+* For each named argument in A, if its corresponding parameter (as described in §7.5.1.1) does not have the same position as that argument, then all the following arguments must be named.
+
+In particular, this prevents invoking `void M(bool a = true, bool b = true, bool c = true, );` with `M(c: false, valueB);`. The first argument is used non-positionally (the argument is used in first position, but the parameter named "c" is in third position), so the following arguments should be named.
+
+In other words, non-trailing named arguments are only allowed when the name and the position result in finding the same corresponding parameter.
+
+## Detailed design (VB)
+[design]: #detailed-design-vb
+
+TODO
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+## Alternatives
+[alternatives]: #alternatives
+
+There are a couple of alternatives to consider:
+
+- The status quo
+- Providing IDE assistance to fill-in all the names of trailing arguments when you type specific a name in the middle.
+
+Both of those suffer from more verbosity, as they introduce multiple named arguments even if you just need one name of a literal at the beginning of the argument list.
+
+## Unresolved questions
+[unresolved]: #unresolved-questions
+
+- implications for F#
```

Comment:
I believe that this would make such code uncallable in F# without changes to the F# compiler to allow non-trailing named parameters.  No idea how much work is involved in the F# side of things.